File: small_projs/promisc/python_pcap/print_headers.py
#!/usr/bin/env python

import sys
from socket import ntohs;

#global_vars
ETHER_LEN = 6

stdwrite = sys.stdout.write;

def print_ether_addr(addr):
	for i in range(ETHER_LEN):
		stdwrite("%2.2X" % addr[i]);
		if i != ETHER_LEN - 1:
			stdwrite(":");

class ether_header:
	def __init__(self, data):
		self.ether_dhost = [ord(addr) for addr in data[0:6]];
		self.ether_shost = [ord(addr) for addr in data[6:12]];
		self.ether_type = ord(data[12]);
	
	def print_dhost(self):
		print_ether_addr(self.ether_dhost);

	def print_shost(self):
		print_ether_addr(self.ether_shost);
	
	def print_type(self):
		stdwrite("%4.4X" % ntohs(self.ether_type));


def print_eth_header(payload):
	hdr = ether_header(payload);
	stdwrite("Dst ");
	hdr.print_dhost();
	stdwrite(" Src ");
	hdr.print_shost();
	stdwrite("\ntype=");
	hdr.print_type();
	stdwrite("\n");


def print_raw_hex_data(payload):
	for i in range(len(payload)):
		if i % 8 == 0 and i != 0:
			stdwrite("\n");
		stdwrite("%2.2X" % ord(payload[i]));
		if i % 2 == 1 and i % 8 != 0:
			stdwrite(" ");
	stdwrite("\n\n");

File: small_projs/promisc/c_pcap/print-headers.c
#include <sys/socket.h>

#include <arpa/inet.h>
#include <netpacket/packet.h>
#include <net/ethernet.h>
#include <stdio.h>
#include <stdint.h>

#include "print-headers.h"

enum {
	ETH_LEN = 6,
};

void print_eth_addr(u_int8_t host[ETH_LEN]);

void
print_main(void *data, int len)
{
	print_eth_header(data);
		
	print_raw_hex_data(data, len);

}

/*
 * FIXME:
 * some strange with pppoe
 * 
 */

int
print_eth_header(void *data)
{
	struct ether_header *ether;
	
	ether = data;
	
	printf("Dst ");
	print_eth_addr(ether->ether_dhost);
	printf(" Src ");
	print_eth_addr(ether->ether_shost);
	
	printf("\ntype=%4.4X", ntohs(ether->ether_type));

	printf("\n");
	
	return sizeof(*ether);
}


void
print_raw_hex_data(void *data, int len)
{
	char *buff;
	int i;
	
	buff = (char *)data;

	for (i = 0; i < len; i++) {
		if (i % 8 == 0 && i != 0)
			printf("\n");
		printf("%2.2x", buff[i] & 0xff);
		if (i % 2 == 1 && i % 8 != 0)
			printf(" ");
	}
}

void
print_eth_addr(u_int8_t host[ETH_LEN])
{
	int i;

	for (i = 0; i < 6; i++) {
		printf("%2.2X", host[i]);
		if (i != ETH_LEN - 1)
			printf(":");
	}

}

